File: backend/app/core/upgrades/stat_modifiers/bullet_damage_2.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class BulletDamage2(UpgradeImplementation):
    id = 'bullet_damage_2'
    name = 'Sharper Rounds'
    description = '+4 bullet damage'
    rarity = 'uncommon'
    type = 'stat_modifier'
    cost = 6

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 4
        appliers.modifiers.add_multiplier('attack', 'damage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/bullet_damage_2.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID, AttackTypeID } from '../../data/ID'

export const BulletDamage2Def: UpgradeDef = {
  id: "bullet_damage_2",
  name: "Sharper Rounds",
  description: "+4 bullet damage",
  rarity: RarityID.Uncommon,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 6,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.Damage,
  value: 4,
  isMultiplier: false,
  specificAttackType: AttackTypeID.Bullet,
  stackable: true,
  maxStacks: 99999,
}

export class BulletDamage2 extends Upgrade {}
